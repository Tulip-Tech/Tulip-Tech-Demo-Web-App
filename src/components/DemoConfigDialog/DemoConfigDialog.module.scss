@use 'src/styles/variables';
@use 'src/styles/theme';

.controls {
  margin-top: variables.$base-spacing;

  > button {
    margin-right: 8px;
  }
}

.maxWidth {
  max-width: 500px;
}

.configModal {
  position: absolute;
  top: 0;
  width: 100%;
  height: 100%;
  background: var(--body-background-color);

  p {
    max-width: 500px;
    margin-bottom: 0;
    color: theme.$text-field-resting-color;
    font-size: 14px;
    line-height: 18px;
  }

  a,
  a:visited,
  a:hover,
  a:active {
    color: inherit;
    font-weight: var(--body-font-weight-bold);
    text-decoration: underline;
    cursor: pointer;
  }
}

.note {
  padding: variables.$base-spacing;
  text-align: center;

  a {
    text-decoration: underline;
    cursor: pointer;
  }

  * {
    margin-top: calc(#{variables.$base-spacing} / 2);
    color: theme.$text-field-resting-color;
    font-weight: var(--body-font-weight-bold);
  }
}
