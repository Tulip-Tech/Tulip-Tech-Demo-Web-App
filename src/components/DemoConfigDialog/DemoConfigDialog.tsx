import React, { ChangeEventHandler, MouseEventHandler, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useNavigate } from 'react-router';
import type { UseQueryResult } from 'react-query';

import styles from './DemoConfigDialog.module.scss';

import ErrorPage from '#components/ErrorPage/ErrorPage';
import TextField from '#components/TextField/TextField';
import Button from '#components/Button/Button';
import { getConfigNavigateCallback } from '#src/utils/configOverride';
import Link from '#components/Link/Link';
import ConfirmationDialog from '#components/ConfirmationDialog/ConfirmationDialog';
import LoadingOverlay from '#components/LoadingOverlay/LoadingOverlay';
import type { Config } from '#types/Config';
import DevStackTrace from '#components/DevStackTrace/DevStackTrace';

const regex = /^[a-z,\d]{0,8}$/g;
const fallbackConfig = import.meta.env.APP_DEMO_FALLBACK_CONFIG_ID;

interface Props {
  selectedConfigSource: string | undefined;
  configQuery: UseQueryResult<Config>;
}

interface State {
  configSource: string | undefined;
  error: string | Error | undefined;
  showDialog: boolean;
  loaded: boolean;
}

const initialState: State = {
  configSource: undefined,
  error: undefined,
  showDialog: false,
  loaded: false,
};

const DemoConfigDialog = ({ selectedConfigSource, configQuery }: Props) => {
  const { t } = useTranslation('demo');
  const navigate = useNavigate();
  const navigateCallback = getConfigNavigateCallback(navigate);

  const [state, setState] = useState<State>(initialState);

  const configNavigate = async (configSource: string | undefined) => {
    setState((s) => ({ ...s, configSource: configSource, error: undefined }));

    if (!configSource) {
      setState((s) => ({ ...s, error: t('enter_a_value') }));
    }
    // If trying to fetch the same config again, use refetch since a query param change won't work
    else if (configSource === selectedConfigSource) {
      await configQuery.refetch();
    }
    // Get a new config by triggering a query param change
    else {
      navigateCallback(configSource);
    }
  };

  useEffect(() => {
    // Don't grab values from props when config source is unset or still loading
    if (!selectedConfigSource || configQuery.isLoading) {
      return;
    }

    // Initialize the config source if it's not yet set (this happens at first load)
    setState((s) => ({ ...s, configSource: s.configSource ?? selectedConfigSource }));

    // If there's an error after loading is done, grab it to display it to the user
    if (configQuery.error) {
      setState((s) => ({ ...s, showDialog: false, error: configQuery.error as Error }));
    }
  }, [selectedConfigSource, configQuery.error, configQuery.isLoading]);

  const clearConfig = () => {
    setState(initialState);
    navigateCallback('');
  };

  const isValidSource = (configSource: string) => configSource.match(regex)?.some((m) => m === configSource);

  const onChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    setState((s) => ({
      ...s,
      configSource: event.target.value,
      error: isValidSource(event.target.value || '') ? undefined : t('invalid_config_id_format'),
    }));
  };

  const submitClick: MouseEventHandler<HTMLButtonElement> = async (event) => {
    event.preventDefault();

    let error: string = '';

    if (state.configSource?.length !== 8) {
      error = t('enter_8_digits');
    } else if (!isValidSource(state.configSource)) {
      error = t('invalid_config_id_format');
    }

    if (error) {
      setState((s) => ({ ...s, error: error }));
    } else {
      await configNavigate(state.configSource);
    }
  };

  const cancelConfigClick: MouseEventHandler<HTMLButtonElement> = (event) => {
    event.preventDefault();

    if (fallbackConfig) {
      setState({ configSource: '', error: undefined, loaded: false, showDialog: true });
    }
  };

  const confirmDemoClick = async () => {
    await configNavigate(fallbackConfig);
  };

  const cancelDemoClick = () => setState((s) => ({ ...s, showDialog: false }));

  // If the config loads, reset the demo ui state
  useEffect(() => {
    if (configQuery.isSuccess) {
      setState(initialState);
    }
  }, [setState, configQuery.isSuccess]);

  // If someone links to the app with a config query,
  // we want to show the normal spinner instead of the dialog while trying to load the first time
  // Config is only undefined when it's the first load attempt.
  if (configQuery.isLoading && state.configSource === undefined) {
    return <LoadingOverlay />;
  }

  return (
    <>
      {configQuery.isSuccess && (
        <div className={styles.note}>
          <div>{t('currently_previewing_config', { configSource: selectedConfigSource })}</div>
          <Link onClick={clearConfig}>{t('click_to_unselect_config')}</Link>
        </div>
      )}
      {!configQuery.isSuccess && (
        <div className={styles.configModal}>
          <ErrorPage
            title={t('app_config_not_found')}
            helpLink={'https://docs.jwplayer.com/platform/docs/ott-create-an-app-config'}
            error={typeof state.error === 'string' ? undefined : state.error}
          >
            <form method={'GET'} target={'/'}>
              <TextField
                required
                disabled={configQuery.isLoading}
                className={styles.maxWidth}
                maxLength={8}
                value={state.configSource || ''}
                placeholder={t('please_enter_config_id')}
                name={'app-config'}
                autoCapitalize={'none'}
                error={!!state.error}
                helperText={
                  state.error && (
                    <span>
                      {typeof state.error === 'string' ? state.error : state.error?.message}
                      <br />
                      <br />
                      <DevStackTrace error={typeof state.error === 'string' ? undefined : state.error} />
                    </span>
                  )
                }
                onChange={onChange}
              />
              <div className={styles.controls}>
                <Button
                  label={t('submit_config_id')}
                  type={'submit'}
                  onClick={submitClick}
                  disabled={configQuery.isLoading}
                  busy={configQuery.isLoading && !state.showDialog}
                />
                {fallbackConfig && <Button label={t('cancel_config_id')} onClick={cancelConfigClick} disabled={configQuery.isLoading} />}
              </div>
            </form>
            <p>{t('use_the_jwp_dashboard')}</p>
            <p>{t('demo_note')}</p>
          </ErrorPage>
        </div>
      )}
      <ConfirmationDialog
        open={state.showDialog}
        title={t('view_generic_demo')}
        body={t('viewing_config_demo')}
        onConfirm={confirmDemoClick}
        onClose={cancelDemoClick}
        busy={configQuery.isLoading}
      />
    </>
  );
};

export default DemoConfigDialog;
