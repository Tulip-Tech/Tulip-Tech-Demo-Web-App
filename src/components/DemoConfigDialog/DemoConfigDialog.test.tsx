import React from 'react';
import type { UseQueryResult } from 'react-query';

import { renderWithRouter } from '#test/testUtils';
import DemoConfigDialog from '#components/DemoConfigDialog/DemoConfigDialog';
import type { Config } from '#types/Config';

describe('<DemoConfigDialog>', () => {
  test('renders and matches snapshot', () => {
    const { container } = renderWithRouter(<DemoConfigDialog configQuery={{ isSuccess: true } as UseQueryResult<Config>} selectedConfigSource={'abcdefgh'} />);

    expect(container).toMatchSnapshot();
  });

  test('renders and matches snapshot error dialog', () => {
    const { container } = renderWithRouter(<DemoConfigDialog configQuery={{ isSuccess: false } as UseQueryResult<Config>} selectedConfigSource={'aaaaaaaa'} />);

    expect(container).toMatchSnapshot();
  });
});
