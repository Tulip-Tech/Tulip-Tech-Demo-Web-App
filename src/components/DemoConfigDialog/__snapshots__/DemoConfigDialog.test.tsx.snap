// Vitest Snapshot v1

exports[`<DemoConfigDialog> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_note_561a54"
  >
    <div>
      currently_previewing_config
    </div>
    <a
      class="_link_a14c64"
    >
      click_to_unselect_config
    </a>
  </div>
</div>
`;

exports[`<DemoConfigDialog> > renders and matches snapshot error dialog 1`] = `
<div>
  <div
    class="_configModal_561a54"
  >
    <div
      class="_errorPage_8c5621"
    >
      <div
        class="_box_8c5621"
      >
        <img
          alt="Logo"
          class="_image_8c5621"
          src="/images/logo.png"
        />
        <header>
          <h1
            class="_title_8c5621"
          >
            app_config_not_found
          </h1>
        </header>
        <main
          class="_main_8c5621"
        >
          <form
            method="GET"
            target="/"
          >
            <div
              class="_textField_e16c1b _maxWidth_561a54"
            >
              <label
                class="_label_e16c1b"
                for="text-field_1235_app-config"
              />
              <div
                class="_container_e16c1b"
              >
                <input
                  autocapitalize="none"
                  class="_input_e16c1b"
                  id="text-field_1235_app-config"
                  maxlength="8"
                  name="app-config"
                  placeholder="please_enter_config_id"
                  required=""
                  type="text"
                  value="aaaaaaaa"
                />
              </div>
            </div>
            <div
              class="_controls_561a54"
            >
              <button
                class="_button_f8f296 _default_f8f296 _outlined_f8f296"
                type="submit"
              >
                <span>
                  submit_config_id
                </span>
              </button>
            </div>
          </form>
          <p>
            use_the_jwp_dashboard
          </p>
          <p>
            demo_note
          </p>
          <p
            class="_links_8c5621"
          >
            <a
              href="https://docs.jwplayer.com/platform/docs/ott-create-an-app-config"
              rel="noreferrer"
              target="_blank"
            >
              learn_more
            </a>
          </p>
        </main>
      </div>
    </div>
  </div>
</div>
`;
